File: minimizeHeightsII.cpp
class Solution {
public:
    int getMinDiff(int arr[], int n, int k) {
        // code here
        if (n== 1){
            return 0;
        }
        int result = INT_MAX;
        vector<int> visited(n, 0);
        vector<pair<int, int>> pairs;
        for (int i = 0; i < n; i++){
            if (arr[i] - k >= 0) pairs.push_back({arr[i] - k, i});
            pairs.push_back({arr[i] + k, i});
        }
        sort(pairs.begin(), pairs.end());
//        for (auto i: pairs){
//            cout << "( "<<i.first << " " << i.second << " )" ;
//        }
        int i = 0;
        int j = 0;
        int size = pairs.size();
        int count = 0;
        while (count < n and j < size){
            if (visited[pairs[j].second] == 0) count++;
            visited[pairs[j].second] += 1;
            j++;
        }
        int ans = pairs[j-1].first - pairs[i].first;
        while (j < size){

            if (visited[pairs[i].second] == 1){
                count--;
            }
            visited[pairs[i].second] -= 1;
            i++;

            while (count < n and j < size){
                if (visited[pairs[j].second] == 0) count++;
                visited[pairs[j].second] += 1;
                j++;
            }
            if (count == n){
                ans = min(ans, pairs[j-1].first - pairs[i].first);
            }


        }
        return ans;

    }
};

File: minimizeHeightsII.py
//time: O(nlogn)
class Pair:
    def __init__(self, value, index):
        self.value = value
        self.index = index
class Solution:
    def getMinDiff(self, arr, n, k):
        if len(arr) == 1 :
            return 0
        visited = []
        possibleHeights = []
        for i in range(n):
            visited.append(0)
            if arr[i] - k >= 0:
                possibleHeights.append(Pair(arr[i] - k, i))
            possibleHeights.append(Pair(arr[i] + k, i))
        possibleHeights = sorted(possibleHeights, key = lambda x: x.value)
        left = 0
        right = 0
        count = 0
        ans = float('inf')
        while right < len(possibleHeights):
            if left + 1 > n:
                break
            visited[possibleHeights[right].index] += 1
            if visited[possibleHeights[right].index] == 1:
                count += 1
            if count == n:
                # check ans until count is less than n
                while count == n:
                    if count == n:
                        ans = min(ans, possibleHeights[right].value - possibleHeights[left].value)
                    visited[possibleHeights[left].index] -= 1
                    if visited[possibleHeights[left].index] == 0:
                        count -= 1
                    left += 1
            right += 1
        return ans
